common/src/resources/range_map.rs

```diff
@@ -0,0 +1,236 @@
+use std::{collections::HashMap, ops::Range};
+
+use crate::ether::evm::core::opcodes::WrappedOpcode;
+
+#[derive(Copy, Clone, Debug)]
+enum CollisionKind {
+    Deletion,
+    Splitting,
+    Shortening,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct RangeMap(pub HashMap<Range<usize>, WrappedOpcode>);
+
+impl RangeMap {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    /// Given an offset into memory, returns the associated opcode if it exists
+    pub fn get_by_offset(&self, offset: usize) -> Option<WrappedOpcode> {
+        self.0.get(self.find_range(offset).expect("RangeMap::have_range is broken")).cloned()
+    }
+
+    /// Associates the provided opcode with the range of memory modified by writing a `size`-byte
+    /// value to `offset`.
+    ///
+    /// This range is exactly `[offset, size - 1]`. This function ensures that any existing ranges
```

Comment:
the range should be 

`[offset, offset + size]`

see https://github.com/Jon-Becker/heimdall-rs/blob/7c7ca04d2f9d6813d880038058be38f76f6c8f6a/common/src/ether/evm/core/memory.rs#L88-L89

Reply:
Good catch, but if I'm interpreting the `Vec::splice` semantics correctly then it's `[offset, offset + size - 1]` (I interpret the `[]` notation as inclusive)?

Reply:
Yep, that makes sense!

Reply:
Consult [d1d508f](https://github.com/Jon-Becker/heimdall-rs/pull/242/commits/d1d508fb732e5c341df2ca16fb4bb9d814e114b5).